src/main/java/com/puppycrawl/tools/checkstyle/checks/design/FinalClassCheck.java

```diff
@@ -256,90 +267,120 @@ private static boolean shouldBeDeclaredAsFinal(ClassDesc desc) {
+     * @return {@link Optional} of {@link ClassDesc} object of the nearest class with the same name.
      * @noinspection CallToStringConcatCanBeReplacedByOperator
      */
-    private ClassDesc getNearestClassWithSameName(String className, String superClassName) {
+    private Optional<ClassDesc> getNearestClassWithSameName(String className,
+        Function<String, Integer> countProvider) {
         final String dotAndClassName = PACKAGE_SEPARATOR.concat(className);
         return innerClasses.entrySet().stream()
                 .filter(entry -> entry.getKey().endsWith(dotAndClassName))
                 .map(Map.Entry::getValue)
                 .min((first, second) -> {
-                    return getClassDeclarationNameMatchingCountDiff(superClassName, first, second);
-                })
-                .orElse(null);
-    }
-
-    /**
-     * Get the difference between class declaration name matching count. If the difference between
-     * them is zero, then their depth is compared to obtain the result.
-     *
-     * @param superClassName name of the super class
-     * @param firstClass first input class
-     * @param secondClass second input class
-     * @return difference between class declaration name matching count
-     */
-    private static int getClassDeclarationNameMatchingCountDiff(String superClassName,
-                                                                ClassDesc firstClass,
-                                                                ClassDesc secondClass) {
-        int diff = Integer.compare(
-            CheckUtil
-                .typeDeclarationNameMatchingCount(superClassName, secondClass.getQualifiedName()),
-            CheckUtil
-                .typeDeclarationNameMatchingCount(superClassName, firstClass.getQualifiedName()));
-        if (diff == 0) {
-            diff = Integer.compare(firstClass.getDepth(), secondClass.getDepth());
-        }
-        return diff;
-    }
-
-    /**
-     * Check if class name matches with anonymous inner class name.
-     *
-     * @param ast current ast.
-     * @param classDesc class to match.
-     * @return true if current class name matches anonymous inner
-     *         class name.
-     */
-    private static boolean doesNameOfClassMatchAnonymousInnerClassName(DetailAST ast,
-                                                               ClassDesc classDesc) {
-        final String[] className = classDesc.getQualifiedName().split("\\.");
-        return ast.getFirstChild().getText().equals(className[className.length - 1]);
+                    int diff = Integer.compare(
+                        countProvider.apply(second.getQualifiedName()),
+                        countProvider.apply(first.getQualifiedName())
+                    );
+                    if (diff == 0) {
+                        diff = Integer.compare(first.getDepth(), second.getDepth());
+                    }
+                    return diff;
+                });
```

Comment:
@strkkk Should I refactor this to-
```
.max(Comparator.comparingInt(
    (ClassDesc classDesc) -> countProvider.apply(classDesc.getQualifiedName()))
    .thenComparingInt(classDesc -> -classDesc.getDepth()));
```
The only drawback that I feel there is that we have to use `-` (minus) sign in the latter comparison. What do you think?

Reply:
I thought about this and I find this way too noisy. As for me, this way would be better if there was possibility to use method refs instead of lambdas in both comparisons.

By the way, this comparison method can be more simplified with
```
final int diff = ...
return diff != 0 ? diff : second.getDepth() - first.getDepth();
```

Reply:
> As for me, this way would be better if there was possibility to use method refs instead of lambdas in both comparisons.

I changed the implementation to use method ref in both comparisions.